### PR TITLE
internal/router: do not always resolve ids

### PR DESCRIPTION
**Pull request by rogpeppe**
When there's a PUT or a POST, we might actually be
creating a new entity, so defer id resolution
in that case so the specific handler can
deal with it.

For POSTs and PUTs to metadata, the id must already
exist, because otherwise there could not be
metadata about it, so we always resolve in that
case.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/67/

File: internal/v4/api_test.go [hunk 1]
```diff
@@ -281,15 +281,14 @@ func (s *APISuite) TestBulkMeta(c *gc.C) {
...
-			},
-		})
+	storetesting.AssertJSONCall(c, storetesting.JSONCallParams{
+		Handler: s.srv,
+		URL:     storeURL("meta/charm-metadata?id=precise/wordpress-23&id=precise/mysql-10"),
+		ExpectBody: map[string]*charm.Meta{
+			"precise/wordpress-23": wordpress.Meta(),
+			"precise/mysql-10":     mysql.Meta(),
+		},
+	})
```
Comment (urosj) [inline internal/v4/api_test.go:291]: Why this change if the patch deals with get/put resolve issues?

File: internal/v4/api_test.go [hunk 2]
```diff
@@ -299,27 +298,26 @@ func (s *APISuite) TestBulkMetaAny(c *gc.C) {
...
-			Handler: s.srv,
-			URL:     storeURL("meta/any?include=charm-metadata&include=charm-config&id=precise/wordpress-23&id=precise/mysql-10"),
-			ExpectBody: map[string]params.MetaAnyResponse{
-				"precise/wordpress-23": {
-					Id: wordpressURL,
-					Meta: map[string]interface{}{
-						"charm-config":   wordpress.Config(),
-						"charm-metadata": wordpress.Meta(),
-					},
+	storetesting.AssertJSONCall(c, storetesting.JSONCallParams{
```
Comment (urosj) [inline internal/v4/api_test.go:301]: Same here. Even if it's just formatting, I would suggest that next time we do not make the changes that are not directly related to the suggested PR. And create another PR for them.

File: internal/router/router.go [hunk 3]
```diff
@@ -167,15 +167,22 @@ func (r *Router) serveIds(w http.ResponseWriter, req *http.Request) error {
...
-	if handler, ok := r.handlers.Id[key]; ok {
+	handler := r.handlers.Id[key]
+	if handler == nil || idHandlerNeedsResolveURL(req) {
+		// If it's not an id handler, it's a meta endpoint, so
+		// we always want a resolved URL. Otherwise we leave the
+		// URL unresolved for cases where the id may validly not
+		// exist (for example when uploading a new charm).
+		if err := r.resolveURL(url); err != nil {
+			// Note: preserve error cause from resolveURL.
+			return errgo.Mask(err, errgo.Any)
```
Comment (urosj) [inline internal/router/router.go:182]: What do we do with it in the API?

File: internal/v4/api_test.go (hunk 1 above)
Comment (rogpeppe) [inline internal/v4/api_test.go:291]: It's a trivial formatting change, a relic of a previous branch.

File: internal/v4/api_test.go (hunk 2 above)
Comment (rogpeppe) [inline internal/v4/api_test.go:301]: OK, will do.

**Comment by jrwren** [comment]
LGTM

**Comment by rogpeppe** [comment]
```
[10:59:46] <urulama> rogpeppe1: LGTM, but go through comments, please
```

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge
